refactor: use C++ stdlib PRNG

grawlinson opened the pull request:
Still a work in process:

- [x] Fix compile issues
- [x] Fix off-by-one errors & incorrect range given to PRNG
  - PRNG expects `a > b`, sometimes it is actually `a < b`
  - PRNG changed from `[a, b)` to `[a, b]`
- [ ] Get replays working

@ahockersten I'd love if you had a look at this next part because I'm not familiar with the replay code at all.

grawlinson committed:
refactor: use C++ stdlib PRNG

Still a work in process, also fails to compile due to some archive
tomfoolery.

Due to PRNG changing from [a, b) to [a, b], there may be some off by one
errors.

ahockersten commented:
I'm not sure I'm that familiar with the replay code unfortunately. :( If you are referring to the compile errors, I can dig into those but I'm not sure I'll do any better than you.

For the "checking the state of the RNG", I have a dumb idea: clone the random number generator (which C++ supports). From the clone, take the next random number (between `min()` and `max()`) and save **that** for the comparison. That way we can still verify that the replay is matching the original.

Although I'm not sure the above is necessary, the code still checks worm position and velocities and that should be more than enough to detect desyncs.

ahockersten commented on src/game/gameEntry.cpp:
```diff
@@ -15,13 +15,16 @@
-	// TODO: Better PRNG seeding
-	gfx.rand.seed(Uint32(std::time(0)));
+	// TODO: Validate PRNG seeding
+  // why do we have *two* PRNGs (gfx & game)?
```
I think I can answer this one. It's a game-dev trick. By having a separate random state for the game compared to the graphics, the game state and the graphics state are decoupled, so you can make changes to the graphics without it affecting the game.

This is useful if you want to e.g. support different graphics settings (higher graphics => more random particles for example). I'm not sure if it's useful to Liero though.

ahockersten commented:
So I've dived into the remaining issues a bit and I think I have a theory. This call has become problematic after we changed to the new PRNG:
`archive(ar, game.rand);`

It's complaining about a missing overload. Which is true, there is no overload that would accept a `std::mt19937`. But why did it work before? I think that there is an overload somewhere for the old code which allows you to get the seed for `gvl::mwc` by just calling it, so the **actual** overload that gets used here is one for `uint32_t` (or other int type - I haven't been able to find said overload though.

But basically, this code needs to save the RNG seed and then it will work. It's the only thing from the RNG that would make any sense to save, and the archiving code can only really save numbers, so it literally seems like it couldn't be referring to anything else.

grawlinson commented:
Yeah, that makes sense!

I came to the same conclusion that we only need to save the PRNG seed. So it looks like we don’t need to make an archive function just for it.

grawlinson committed:
fix: segfault #1, more to come

grawlinson commented:
Getting segfaults at this point:

```
Thread 1 "openliero" received signal SIGSEGV, Segmentation fault.
0x00005555555bc05a in std::uniform_int_distribution<int>::param_type::b (this=<error reading variable: Cannot access memory at address 0x7fffff7feff8>) at /usr/include/c++/14.2.1/bits/uniform_int_dist.h:116
116             b() const
```

Odds are it'll be something stupid like uniform_int_distribution() getting (0, -1) or (0, undefined). Trying to narrow it down atm.

grawlinson commented:
Found a potential segfault here:

https://github.com/openliero/openliero/blob/ed0228bb7bd073b93abefe7294d47d295360ff05/src/game/worm.cpp#L380

For some reason, the variable `health` occasionally dips into the negatives, thus causing `uniform_int_distribution` to segfault.

Gonna keep checking the other calls to `uniform_int_distribution` for potential issues.

grawlinson committed:
fix: segfault #2, more to come

grawlinson committed:
fix: segfault #3, more to come

grawlinson committed:
refactor: use mt19937 as Game{}'s rand

grawlinson committed:
docs: add another TODO

grawlinson committed:
Merge branch 'master' into use-cxx-random

grawlinson committed:
refactor: fix final compile failure (PRNG seed)

grawlinson committed:
refactor: use rand_seed in fastGameChecksum()

grawlinson commented:
Code compiles, doesn't segfault ... however, replays don't work. Fairly sure the replay code doesn't know how to obtain the PRNG seed from c4b649810b7eb6e3a55710d5028e9db95d15ecb0.

grawlinson committed:
refactor: remove unused windows.hpp

grawlinson committed:
build: update vcpkg baseline

grawlinson committed:
style: spaces to tabs

grawlinson committed:
build: use LTO for Release builds only

grawlinson committed:
refactor: bunch of improvements

* add library (src/game)
* removed include_directories() hack

grawlinson committed:
refactor: fix compile time failures

grawlinson committed:
build: split out tools

grawlinson committed:
refactor: rename *tool's main.cpp

grawlinson committed:
refactor: remove tracing code (redo later)

grawlinson committed:
refactor: remove unnecessary #if 1

grawlinson committed:
build: enable installation of debug binaries

grawlinson committed:
refactor: more gvl pruning

grawlinson committed:
refactor: more gvl pruning #2

grawlinson committed:
refactor: more gvl pruning #3

grawlinson committed:
refactor: GVL_STATIC_ASSERT isn't used anywhere

grawlinson committed:
refactor: what kind of savage disables sound?

grawlinson committed:
refactor: nice knowing you __BORLANDC__

grawlinson committed:
refactor: move around some types

Source of type_info.hpp is here:

https://www.boost.org/doc/libs/1_86_0/boost/python/type_id.hpp

grawlinson committed:
refactor: use stdlib default_delete

grawlinson committed:
refactor: move functions from gvl bits to where they're used

grawlinson committed:
build: organise CMake presets alphabetically

grawlinson committed:
refactor: add Weapon::ShotType enum

grawlinson committed:
refactor: use switch-case for Weapon::ShotType

grawlinson committed:
style: throw gvl through clang-format

grawlinson committed:
style: throw {tc,video}_tool through clang-format

grawlinson committed:
style: throw game through clang-format

grawlinson committed:
docs: start documenting Worm/NinjaRope

grawlinson committed:
refactor: start using enum class for Reaction Force

... as well as adding some more Worm documentation.

grawlinson committed:
refactor: switch type-casting from C to C++

From (type)value to static_cast<type>(value).

grawlinson committed:
refactor: rename gvl headers for accuracy

grawlinson committed:
refactor: start using enum class for Worm Controls

WormSettings::Value -> WormSettings::Control::Value

grawlinson committed:
refactor: start using enum class for Worm Controls

Worm::Value -> Worm::Control::Value

grawlinson committed:
refactor: move Worm struct initialisation values around

grawlinson committed:
feat: add enum for Worm direction

grawlinson committed:
refactor: remove unused objectList.hpp

grawlinson committed:
refactor: explicitly use Settings::GameMode::Value enum

grawlinson committed:
fix: cppcheck's missingOverride warnings

grawlinson committed:
fix: cppcheck's const* warnings

* constVariablePointer
* constVariableReference
* constParameterReference
* constParameterPointer

grawlinson committed:
fix: cppcheck's passedByValue warnings

grawlinson committed:
fix: FTBFS for WIN32

grawlinson committed:
fix: cppcheck's uselessOverride warning

grawlinson committed:
fix: cppcheck's cStyleCast warnings

grawlinson committed:
fix: some of cppcheck's shadow* warnings

grawlinson committed:
fix: some of cppcheck's unreadVariable warnings

grawlinson committed:
fix: cppcheck's returnStdMoveLocal warnings

grawlinson committed:
fix: some of cppcheck's unreadVariable warnings

grawlinson committed:
fix: replace localtime with reentrant implementation

grawlinson committed:
fix: FTBFS with localtime() and MSVC

grawlinson committed:
fix: some more cppcheck warnings

grawlinson committed:
style: shuffle around comment

grawlinson committed:
style: no need for 'using' for a couple of instances

grawlinson committed:
fix: add missing header <algorithm>

For algo stuff introduced in b2f09c0432cb03a3e884a7d08ffea9bc1a7a952d

grawlinson committed:
refactor: comments/dead code

grawlinson committed:
style: reshuffle headers

grawlinson committed:
fix: explicitly require MSVC 8.0+ for localtime()

grawlinson committed:
refactor: assorted bits and bobs

grawlinson committed:
refactor: split out main.cpp

grawlinson committed:
fix: cppcheck's useStlAlgorithm warnings

grawlinson committed:
fix: remove negative comparison against unsigned

grawlinson committed:
refactor: flesh out menu enums

grawlinson committed:
style: shuffle comments around

grawlinson committed:
fix: cppcheck's useInitializationList warnings

grawlinson committed:
fix: sound regression

Fixes regression introduced in b2f09c0432cb03a3e884a7d08ffea9bc1a7a952d
with std::fill.

grawlinson committed:
refactor: change display string for Holdazone

From 'Holdazone' to 'Hold a Zone'.

grawlinson committed:
style: move comments

grawlinson committed:
fix: ensure sound samples are not zeroed when copying

grawlinson committed:
style: remove inconsistent usage of 'using'

grawlinson committed:
refactor: pass time_t as const& to localtime_reentrant()

grawlinson committed:
refactor: delete unused work_queue.hpp

grawlinson committed:
build: add example cppcheck bash script

grawlinson committed:
build: add log files to gitignore

grawlinson committed:
refactor: remove more unused code from gvl

grawlinson committed:
refactor: remove more unused code from game
